fix: Add support for list-type JSON Schema fields in modeling.py

sakthikannan25 opened the pull request:
Problem:
- JSON Schema allows specifying multiple types using array notation: {"type": ["string", "number"]}
- This is valid per JSON Schema specification, and common in real-world schemas
- The mcpadapt modeling.py module failed with "unhashable type: 'list'" when processing such schemas
- Error occurs in get_field_type() when attempting to use a list as a dictionary key

Solution:
- Enhanced get_field_type() to properly handle list-type JSON Schema types
- Added special case to detect when json_type is a list
- Implemented conversion of list-type to Python Union types
- For single-item lists, extract and use the single type
- For multi-item lists, create a Union of all mapped types
- Preserves original behavior for all other schema types

This fix ensures compatibility with JSON Schema that use the array notation for specifying multiple allowed types for a field, which is a common pattern in the JSON Schema ecosystem. The fix is backwards compatible and follows the expected behavior of properly converting JSON Schema types to their Python equivalents.

grll commented:
Hey @sakthikannan25 thanks for the PR I will review soon!

grll commented on tests/test_modeling.py:
```diff
@@ -0,0 +1,223 @@
+        print(f"Error during schema inspection: {e}", file=sys.stderr)
+
+
+def test_direct_modeling_with_list_type():
```
this test could be nice but I would remove the verbose try / except explaining the type error and just keep it to the essential

grll commented on tests/test_modeling.py:
I would keep only 2 tests 1 more end to end as you have and move it to `test_langchain_adapter.py` the other one that simply test the modeling with a schema and test the output ok to keep here but I would move the file in `tests/utils/test_modeling.py`.

grll reviewed (changes requested):
Sorry for your patience, this change look good just a few shuffling around and remove of some verbosity in the tests. Thanks again for your contribution!

sakthikannan25 commented:
Thank you for your feedback on the PR. I've implemented your suggested changes and also found that the same JSON Schema array notation issue needed fixing in the langchain_adapter.py file as well.

Changes made:
1. Fixed modeling.py to handle JSON Schema list-type fields as requested
2. Extended the same fix to langchain_adapter.py to ensure consistent behavior
   - Added support for both array notation (`"type": ["string", "number"]`) and anyOf structures
   - Implemented proper type mapping to Python's Union types
   - Added fallback handling when type information is missing or in an unexpected format
3. Reorganized tests as suggested:
   - Moved a simplified direct test to tests/utils/test_modeling.py
   - Added the E2E test to test_langchain_adapter.py with LangChain-specific adaptations
   - Removed the redundant test file

The tests now pass and demonstrate that our fix properly handles JSON Schema array notation in both modules. I also added some notes in the LangChain test about potential future improvements for nullable field handling, but that would be beyond the scope of this PR.

Let me know if you'd like any additional changes!

grll commented:
> Thank you for your feedback on the PR. I've implemented your suggested changes and also found that the same JSON Schema array notation issue needed fixing in the langchain_adapter.py file as well.
> 
> Changes made:
> 
> 1. Fixed modeling.py to handle JSON Schema list-type fields as requested
> 2. Extended the same fix to langchain_adapter.py to ensure consistent behavior
>    
>    * Added support for both array notation (`"type": ["string", "number"]`) and anyOf structures
>    * Implemented proper type mapping to Python's Union types
>    * Added fallback handling when type information is missing or in an unexpected format
> 3. Reorganized tests as suggested:
>    
>    * Moved a simplified direct test to tests/utils/test_modeling.py
>    * Added the E2E test to test_langchain_adapter.py with LangChain-specific adaptations
>    * Removed the redundant test file
> 
> The tests now pass and demonstrate that our fix properly handles JSON Schema array notation in both modules. I also added some notes in the LangChain test about potential future improvements for nullable field handling, but that would be beyond the scope of this PR.
> 
> Let me know if you'd like any additional changes!

Thanks for the change I will have a look asap!

grll reviewed (approved):
looking good thanks for the change and the contribution!
